zebra-state/src/service/check/tests/prop.rs

```diff
@@ -0,0 +1,411 @@
+        joinsplit2.nullifiers[0] = duplicate_nullifier;
+
+        // make sure there are no other nullifiers
+        joinsplit_data1.first = joinsplit1.0;
+        joinsplit_data1.rest = Vec::new();
+
+        joinsplit_data2.first = joinsplit2.0;
+        joinsplit_data2.rest = Vec::new();
+
+        let transaction1 = transaction_v4_with_joinsplit_data(joinsplit_data1.0);
+        let transaction2 = transaction_v4_with_joinsplit_data(joinsplit_data2.0);
+
+        block1.transactions[0] = transaction_v4_from_coinbase(&block1.transactions[0]).into();
+        block2.transactions[0] = transaction_v4_from_coinbase(&block2.transactions[0]).into();
+
+        block1
+            .transactions
+            .push(transaction1.into());
+        block2
+            .transactions
+            .push(transaction2.into());
+
+        let (mut state, _genesis) = new_state_with_mainnet_genesis();
+
+        let block1_hash;
+        // randomly choose to commit the next block to the finalized or non-finalized state
+        if duplicate_in_finalized_state {
+            let block1 = FinalizedBlock::from(Arc::new(block1));
+            let commit_result = state.disk.commit_finalized_direct(block1.clone(), "test");
+
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+            prop_assert!(commit_result.is_ok());
+
+            block1_hash = block1.hash;
+        } else {
+            let block1 = Arc::new(block1).prepare();
+            let commit_result =
+                state.validate_and_commit(block1.clone());
+
+            prop_assert_eq!(commit_result, Ok(()));
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+
+            block1_hash = block1.hash;
+        }
+
+        let block2 = Arc::new(block2).prepare();
+        let commit_result =
+            state.validate_and_commit(block2);
+
+        prop_assert_eq!(
+            commit_result,
+            Err(
+                DuplicateSproutNullifier {
+                    nullifier: duplicate_nullifier,
+                    in_finalized_state: duplicate_in_finalized_state,
+                }.into()
+            )
+        );
+        prop_assert_eq!(Some((Height(1), block1_hash)), state.best_tip());
+    }
+}
+
+/// Return a new `StateService` containing the mainnet genesis block.
+/// Also returns the finalized genesis block itself.
+fn new_state_with_mainnet_genesis() -> (StateService, FinalizedBlock) {
+    let genesis = zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES
+        .zcash_deserialize_into::<Arc<Block>>()
+        .expect("block should deserialize");
+
+    let mut state = StateService::new(Config::ephemeral(), Mainnet);
+
+    assert_eq!(None, state.best_tip());
+
+    let genesis = FinalizedBlock::from(genesis);
+    let commit_result = state.disk.commit_finalized_direct(genesis.clone(), "test");
+
+    assert_eq!(Some((Height(0), genesis.hash)), state.best_tip());
+    assert!(commit_result.is_ok());
```

Comment:
Fixed in cdb1b80a.

I couldn't fix it before I moved the code, because it was using `prop_assert!`. But there's no random data here, so it should consistently succeed.